### ISSUE TITLE
Cannot invoke "org.nd4j.linalg.api.ndarray.INDArray.detach()" because "fwdPass.lastAct" is null

### ISSUE THREAD

**Issue by sascha08-15** [deeplearning4j/deeplearning4j#9369]
After updating from beta7 to 1.0.0-M1, I get the following exception using the early stopping:
````
21:51:40.506 [WARN ] [BaseEarlyStoppingTrainer fit]: Early stopping training terminated due to exception at epoch 0, iteration 0
java.lang.NullPointerException: Cannot invoke "org.nd4j.linalg.api.ndarray.INDArray.detach()" because "fwdPass.lastAct" is null
	at org.deeplearning4j.nn.layers.recurrent.LSTM.rnnActivateUsingStoredState(LSTM.java:211)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.ffToLayerActivationsInWs(MultiLayerNetwork.java:1139)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore(MultiLayerNetwork.java:2781)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore(MultiLayerNetwork.java:2739)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:174)
21:51:41.298 [INFO ] [Training earlyStoppingTraining]: Training for symbols [NVDA-US] completed
21:51:41.299 [INFO ] [Training earlyStoppingTraining]: Termination reason: Error
21:51:41.299 [INFO ] [Training earlyStoppingTraining]: Termination details: java.lang.NullPointerException: Cannot invoke "org.nd4j.linalg.api.ndarray.INDArray.detach()" because "fwdPass.lastAct" is null
````
Any idea?

#### Version Information

Please indicate relevant versions, including, if relevant:

* Deeplearning4j version: 1.0.0-M1
* Platform information (OS, etc): Win 10 (latest) and Ubunutu 20.10
* No cuda issue

**Comment by agibsonccc** [comment]
@sascha08-15 can you give me something that reproduces this? A standalone example I can copy and paste and just run would get this fixed (or at least verified and a  workaround given) quick. I would also try to run this with snapshots as well.

**Comment by glassbell** [comment]
Getting a similar exception while trying to train a network in RL4J.
```
java.lang.NullPointerException: Cannot invoke "org.nd4j.linalg.api.ndarray.INDArray.detach()" because "fwdPass.lastAct" is null
	at org.deeplearning4j.nn.layers.recurrent.LSTM.rnnTimeStep(LSTM.java:195)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.outputOfLayerDetached(MultiLayerNetwork.java:1342)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.rnnTimeStep(MultiLayerNetwork.java:3179)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.rnnTimeStep(MultiLayerNetwork.java:3160)
	at org.deeplearning4j.rl4j.network.ac.ActorCriticSeparate.output(ActorCriticSeparate.java:250)
	at org.deeplearning4j.rl4j.policy.ACPolicy.nextAction(ACPolicy.java:76)
	at org.deeplearning4j.rl4j.policy.ACPolicy.nextAction(ACPolicy.java:39)
	at org.deeplearning4j.rl4j.learning.async.AsyncThreadDiscrete.trainSubEpoch(AsyncThreadDiscrete.java:116)
	at org.deeplearning4j.rl4j.learning.async.AsyncThread.handleTraining(AsyncThread.java:188)
	at org.deeplearning4j.rl4j.learning.async.AsyncThread.run(AsyncThread.java:164)
```
Not sure how to make a reproduceable example, but it happens when I'm setting `.useLSTM(true)`  like
```
ActorCriticFactorySeparateStdDense.Configuration configuration = ActorCriticFactorySeparateStdDense.Configuration
                .builder().updater(new Adam(0.0003)).useLSTM(true).l2(0).numHiddenNodes(5).numLayer(5).build();
```
If LSTM is not used, everything runs smoothly.
Also I had to shape the input, which is a vector of 5, into an [1, 5, 1] matrix, not sure if it's how it is supposed to be used...

Appears in DL4J version 1.0.0-M1.

**Comment by agibsonccc** [comment]
@FrBst  awesome thanks for the reproducer, let me see if I can come up with a quick fix for this. I suspect it will be fairly easy to get this before the minor release.

**Comment by agibsonccc** [comment]
@FrBst  could you tell me a bit about how you're running everything? I still can't reproduce that locally. I'm running like this (the equivalent configuration from underneath rl4j):
```java
 NeuralNetConfiguration.ListBuilder confB = new NeuralNetConfiguration.Builder().seed(1234)
                .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
                .updater(new Adam())
                .weightInit(WeightInit.XAVIER)
                .l2(1e-1)
                .list().layer(0, new DenseLayer.Builder().nIn(5).nOut(5)
                        .activation(Activation.RELU).build())
                .layer(1, new LSTM.Builder().nOut(6).activation(Activation.TANH).build())
                .layer(2, new RnnOutputLayer.Builder(LossFunctions.LossFunction.MSE).activation(Activation.IDENTITY)
                        .nIn(5).nOut(1).build());
        confB.setInputType(InputType.recurrent(5));
        MultiLayerNetwork multiLayerNetwork = new MultiLayerNetwork(confB.build());
        multiLayerNetwork.init();
        multiLayerNetwork.output(Nd4j.ones(1,5,1));
   ```     

Are there multiple threads running in your setup?
@sascha08-15  if you ever see this, I would also appreciate your reproducer as well.

**Comment by glassbell** [comment]
@agibsonccc 
I have a custom environment, it gives the network input like this:
`return Nd4j.create(new double[][][]{{{cash}, {stocksCount}, {avgPrice}, {currentPrice}, {currentGlobalMetric}}});`

And then I do the following:
```
A3CDiscrete.A3CConfiguration A3C =
                A3CDiscreteDense.A3CConfiguration.builder()
                        .seed(5)
                        .maxEpochStep(365 * 11)
                        .maxStep(365 * 11 * maxTries)
                        .numThread(8)
                        .nstep(10)
                        .updateStart(0)
                        .rewardFactor(0.1)
                        .gamma(0.9999)
                        .errorClamp(1.0)
                        .build();

ActorCriticFactorySeparateStdDense.Configuration configuration = ActorCriticFactorySeparateStdDense.Configuration
                .builder().updater(new Adam(0.0003)).useLSTM(true).l2(0).numHiddenNodes(5).numLayer(5).build();

MDP<SimpleBroker, Integer, DiscreteSpace> mdp = new BrokerMdp(5000, new Random(3));

A3CDiscreteDense<SimpleBroker> a3c = new A3CDiscreteDense<SimpleBroker>(mdp, configuration, A3C);

a3c.train();
```
> Are there multiple threads running in your setup?

I'm using RL4J's A3C with 8 threads.

**Comment by glassbell** [comment]
Tried running https://github.com/treo/rl4j-snake, but with the input of zeroes and with LSTM in the network, and it's running fine. So I'm going to look if I did something wrong in my program and write here again if the problem is gone.

_edit_: Just pasted my classes into treo's project, and everything is working! Seems like the problem is in my original project's settings or something, I'll try to locate it now.

_edit 2_: Oh well. The problem is indeed the version. 1.0.0-beta7 works, 1.0.0-M1 doesn't.

**Comment by agibsonccc** [comment]
@FrBst  thanks for the confirmation, let me switch the version to M1 on here and reproduce this to get a fix going for you.

**Comment by agibsonccc** [comment]
@FrBst  in your project there (or in another self contained project) do you mind creating a small repo for me to clone? I poked at paul's repo this morning and I'm not quite sure where you made the changes. An exact demo class where I can just inspect the stack trace will help me focus on the fix rather than configuring the project. Appreciate your help here a ton!

**Comment by glassbell** [comment]
@agibsonccc here you go: https://github.com/FrBst/rl4j-snake
Launch TrainA3C's main.

**Comment by agibsonccc** [comment]
@FrBst  this is wonderful thank you. I was able to immediately run it and see the issue. I'll follow up soon.

**Comment by agibsonccc** [comment]
@FrBst @sascha08-15 closing this. The issue is related to a broken code path in MKLDNNLSTMHelper which is already disabled by default in snapshots.  If you need the fix, just go ahead and use snapshots for now.

 https://deeplearning4j.konduit.ai/v/en-1.0.0-m1/multi-project/explanation/snapshots#useful-maven-commands-for-snapshots

The new way to use helpers will be through c++ and configuring a maven classifier. Thanks for helping me confirm this and letting me investigate the core issue.
